x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/trace_waterfall_context.tsx

```diff
@@ -215,3 +255,15 @@ export function groupByParent(items: TraceWaterfallItem[]) {
     return acc;
   }, {});
 }
+
+export function filterMapByCriticalPath(
+  map: Record<string, TraceWaterfallItem[]>,
+  criticalPathSegmentsById: Record<string, CriticalPathSegment<TraceWaterfallItem>[]>
+): Record<string, TraceWaterfallItem[]> {
+  return Object.fromEntries(
```

Comment:
There are intermediate allocations with `fromEntries` and `entries` that could be avoided with

```tsx
const result: Record<string, TraceWaterfallItem[]> = {};
for (const parentId in map) {
  result[parentId] = map[parentId].filter(
    (child) => criticalPathSegmentsById[child.id]?.length
  );
}
```

I'd also suggest looking into the nested loop caused by `filter` here. Perhaps if we precompute the valid child ids we could access the valid child with an O(1) operation